### PR TITLE
More stable algorithm for variance, standard deviation

### LINKED ISSUE #386: use more stable var, std algo
https://www.johndcook.com/blog/2008/09/26/comparing-three-methods-of-computing-standard-deviation/
https://www.johndcook.com/blog/2008/09/28/theoretical-explanation-for-numerical-results/

will require upstream fixes to npg too.

This *should* let us delete this `SKIPPED_FUNCS` bit:
https://github.com/xarray-contrib/flox/blob/4dbadaed6d4fd5b525f017182ce19c13cebb1e24/tests/strategies.py#L35-L39

### LINKED ISSUE #422: Suggested change to std/var preprocessing to improve precision
Hi,
I've noticed that in a few rare situations, groupby and flox can return quite noisy standard deviations. In situations where the mean of an array is much larger than the standard deviation (such as deep ocean salinity, raised [here]()), flox returns noisier values on dask arrays than on loaded numpy arrays. In extreme situations, the standard deviation of a dask array can contain NaNs from square-rooting negative variances. 

I'm guessing it's the same idea as #386, in which case @dcherian has thought about this for much longer than I have. I've done a little bit of looking through the code, and could easily have missed something about how this works with neighbouring functions, but my thoughts on the potential problem and how it might be addressed are below.

Minimal complete verifiable example:
```python
import numpy as np
import xarray as xr

l =12000
np.random.seed(1)
test_data = xr.DataArray(np.random.uniform(0,1,l)/100+1000000,dims=('time',)  # huge mean with relatively small variability
                        ).assign_coords({'month':xr.DataArray(np.arange(l)%12,dims=('time',))})

# with numpy arrays returns reasonable and consistent values
test_data.groupby('month').std('time')
# array([0.00283648, 0.00281895, 0.00287791, 0.00287652, 0.00287337,
#        0.00287037, 0.00289802, 0.00289441, 0.00285839, 0.00296478,
#        0.00284787, 0.00292089])

# using lazy computation/dask
dask_test_data = test_data.chunk({'time':100})
dask_test_data.groupby('month').std('time').load()
# array([0.01118034, 0.01118034, 0.01118034, 0.01581139, 0.        ,
#        0.01581139, 0.01118034, 0.01118034, 0.01118034,        nan,
#               nan, 0.        ])
```

A functional workaround is to subtract the mean before calculating standard deviation:
```python
(dask_test_data.groupby('month')-dask_test_data.groupby('month').mean('time')).groupby('month').std('time').load()
```

My understanding is that the distinction comes from aggregate_npg.py improving precision by [subtracting the first non-nan element of the array](https://github.com/xarray-contrib/flox/blob/ca576812e78b3978421eace6e9dde5a76729ebcc/flox/aggregate_npg.py#L112), a preprocessing step [skipped by aggregations.py](https://github.com/xarray-contrib/flox/blob/ca576812e78b3978421eace6e9dde5a76729ebcc/flox/aggregations.py#L379). This solution is probably not quite as stable as subtracting the mean, but the first element should be really close to the mean if the standard deviation is small, and it might be faster.

I’d suggest that to improve precision and match the numpy engine behaviour in aggregations_npg.py, the flox engine implementation for dask arrays of nanstd,nanvar,std,var could have a preprocessor that looks something like this:

```python
def var_std_preprocess(array, axis): # Not sure of naming conventions, sorry
    """Subtracts first value of array from whole array, 
    to improve numerical precision of nanstd, nanvar, std, var

    Adapted from from argreduce_preprocess and _var_std_wrapper in aggregate_npg.py
    """
    import dask.array  # Copied from argreduce_preprocess, but maybe these shouldn’t be within the function? 
    import numpy as np # For either this function or argreduce_preprocess?
    
    # NEXT LINE IS PSEUDOCODE; I’m not entirely sure how to apply it lazily
    # If it doesn’t cost anything speed wise, then probably better to use mean. Happy to run some time tests on either
    first_elements = nanfirst(array,axis) 

    def subtract_first(array_, first_elements_):
        return array_-first_elements_

    return dask.array.map_blocks(
        subtract_first,
        array,
        first,
        dtype=array.dtype,
        meta=array._meta,
        name="groupby-var_std-preprocess",
    )
```

and is included in the Aggregations definition like so:
```python
nanstd = Aggregation(
    "nanstd",
    preprocess=var_std_preprocess, #UPDATED LINE
    chunk=("nansum_of_squares", "nansum", "nanlen"),
    combine=("sum", "sum", "sum"),
    finalize=_std_finalize,
    fill_value=0,
    final_fill_value=np.nan,
    dtypes=(None, None, np.intp),
    final_dtype=np.floating,
)
```

It seems to work if `first_elements` is naively `array[0]` in the one-dimensional, no-nans case, but I’m not sure how to generalise it and apply nanfirst without the usual layers/wrappers. (aggregate_npg.py uses `first = _get_aggregate(engine).aggregate(group_idx, array, func="nanfirst", axis=axis)`, but I don't think this syntax translates to the flox/dask implementation) . If you can give me a few tips or examples to work from, then I’m happy to try implement this behaviour.

Happy also to discuss alternatives, or to provide a pull request if that's easier to work with.

This is also my first time reading through flox code in detail (it's really nicely written and documented, by the way, was lovely to read), and one of the first times I’ve interacted with public github repos, so I’d appreciate any feedback or corrections on what's useful to provide when describing issues.

### PR DESCRIPTION
**Pull request by jemmajeffree**
Updated algorithm for nanvar, to use an adapted version of the Schubert and Gertz (2018) paper mentioned in #386, following discussion in #422

Closes #386
Closes #422

### CODE REVIEW

File: flox/aggregations.py [hunk 1]
```diff
@@ -343,12 +343,61 @@ def _mean_finalize(sum_, count):
...
 
 
+def _var_combine(array, axis, keepdims=True):
+    def clip_last(array):
+        """Return array except the last element along axis
+        Purely included to tidy up the adj_terms line
+        """
+        not_last = [slice(None, None) for i in range(array.ndim)]
+        not_last[axis[0]] = slice(None, -1)
+        return array[*not_last]
```
Comment (dcherian) [inline flox/aggregations.py:353]: Use `array[..., :-1]` and `array[1:, ...]` instead

File: flox/aggregations.py [hunk 2]
```diff
@@ -343,12 +343,61 @@ def _mean_finalize(sum_, count):
...
+        """Return array except the first element along axis
+        Purely included to tidy up the adj_terms line
+        """
+        not_first = [slice(None, None) for i in range(array.ndim)]
+        not_first[axis[0]] = slice(1, None)
+        return array[*not_first]
+
+    assert len(axis) == 1, "Assuming that the combine function is only in one direction at once"
+
+    # Does this double our memory footprint or are they just views?
```
Comment (dcherian) [inline flox/aggregations.py:365]: yes, these are references, no extra memory needed

File: flox/aggregate_flox.py [hunk 3]
```diff
@@ -251,6 +344,43 @@ def nanmean(group_idx, array, *, axis=-1, size=None, fill_value=None, dtype=None
...
+        axis=axis,
+        size=size,
+        fill_value=fill_value,
+        dtype=dtype,
+    )
+
+    # Calculate sum squared deviations - the main part of variance sum
+    array_means = (
+        array_sums / array_lens
+    )  # Does this risk being run eagerly because it's not wrapped in anything?
```
Comment (dcherian) [inline flox/aggregate_flox.py:370]: No at this point, you have in-memory arrays. It is only run at compute-time.

File: flox/aggregations.py [hunk 4]
```diff
@@ -343,12 +343,61 @@ def _mean_finalize(sum_, count):
...
+    # If there's a huge memory impact, probably better to copy paste array.arrays[1]
+    # in and accept the hit to readability
+    sum_deviations = array.arrays[0]
+    sum_X = array.arrays[1]
+    sum_len = array.arrays[2]
```
Comment (dcherian) [inline flox/aggregations.py:370]: One minor style comment: you can use "tuple unpacking" here
 ```suggestion
    sum_deviations, sum_X, sum_len = array.arrays
```

File: flox/aggregate_flox.py [hunk 5]
```diff
@@ -5,6 +5,99 @@
...
 
+MULTIARRAY_HANDLED_FUNCTIONS = {}
+
+
+class MultiArray:
+    arrays: tuple[np.ndarray, ...]
+
+    def __init__(self, arrays):
+        self.arrays = arrays  # something else needed here to be more careful about types (not sure what)
+        # Do we want to co-erce arrays into a tuple and make sure it's immutable? Do we want it to be immutable?
```
Comment (dcherian) [inline flox/aggregate_flox.py:16]: this is fine as-is

File: flox/aggregations.py [hunk 6]
```diff
@@ -343,12 +343,106 @@ def _mean_finalize(sum_, count):
 )
 
 
+def var_chunk(group_idx, array, *, engine: str, axis=-1, size=None, fill_value=None, dtype=None):
```
Comment (dcherian) [inline flox/aggregations.py:346]: I moved this here, so that we can generalize to "all" engines

File: flox/core.py [hunk 7]
```diff
@@ -1251,7 +1252,8 @@ def chunk_reduce(
     # optimize that out.
     previous_reduction: T_Func = ""
     for reduction, fv, kw, dt in zip(funcs, fill_values, kwargss, dtypes):
-        if empty:
+        # UGLY! but this is because the `var` breaks our design assumptions
+        if empty and reduction is not var_chunk:
```
Comment (dcherian) [inline flox/core.py:1256]: this code path is an "optimization" for chunks that don't contain any valid groups. so `group_idx` is all `-1`.

We will need to override `full` in MultiArray.

File: flox/_version.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__version__ = "0.1.dev657+g619a390.d20250606"
```
[the remark this answers is not in the document]
Comment (jemmajeffree) [reply flox/_version.py]: oops. thanks

File: flox/aggregations.py [hunk 9]
```diff
@@ -343,12 +343,106 @@ def _mean_finalize(sum_, count):
...
+        axis=axis,
+        size=size,
+        fill_value=fill_value,
+        dtype=dtype,
+    )
+
+    array_sums = generic_aggregate(
+        group_idx,
+        array,
+        func="nansum",
```
[the remark this answers is not in the document]
Comment (jemmajeffree) [reply flox/aggregations.py:364]: My first thought is to pass through some kind of "are NaNs okay" boolean variable through to var_chunk and var_combine. Is this what xarray's `skipna` does? Or I think I've seen it done as a string `"propogate"` or `"ignore"`? And then to call the var_chunk and var_combine as a partial.

[the remark this answers is not in the document]
Comment (dcherian) [reply flox/aggregations.py:364]: yes the way I do this in flox is create a `var_chunk = partial(_var_chunk, skipna=False)` and `_nanvar_chunk=partial(_var_chunk, skipna=True)` you can stick this in the Aggregation constructor I think

**Comment by dcherian** [comment]
> Do you think it likely that MultiArray would ever be used for anything else? 

Possibly, but only within flox.

> so the line between "expected behaviour" and "something's not right here" can be more clearly defined.

We can liberally make use of comments, `assert` statements, and `NotImplementedError` exceptions to document choices you have made for `var` specifically. We do _not_ have to build the most general version now


> By "add a new test to test_core.py with your reproducer (though modified to work with pure numpy arrays)", do you mean add the failing code from my original issue to the end of test_core.py, along the lines of:

Yes, a simple-ish one would be fine. `groupby_reduce_all` is quite the monster. More recently, I am using [property tests](https://github.com/xarray-contrib/flox/blob/8cfd999cf3d63a1f375e0447aaacb40ecacbcf9c/tests/test_properties.py#L91) to boost coverage; for which `var` and `std` are now skipped: https://github.com/xarray-contrib/flox/blob/8cfd999cf3d63a1f375e0447aaacb40ecacbcf9c/tests/strategies.py#L111

File: tests/test_core.py [hunk 10]
```diff
@@ -2240,3 +2240,25 @@ def test_sparse_nan_fill_value_reductions(chunks, fill_value, shape, func):
...
+    
+    # These two need to be the same function, but with the offset added and not added
+    no_offset, _ = groupby_reduce(array, labels, engine=engine, func=func)
+    with_offset, _ = groupby_reduce(array+offset, labels, engine=engine, func=func)
+    
+    tol = {"rtol": 1e-8, "atol": 1e-10} # Not sure how stringent to be here
+    
+    # Failure threshold in my external tests is dependent on dask chunksize, maybe needs exploring better?
+    
+    assert_equal(no_offset, with_offset, tol)
```
Comment (jemmajeffree) [inline tests/test_core.py]: Is something like this what you mean here?
> add a new test to test_core.py with your reproducer (though modified to work with pure numpy arrays);

Because of this bit in aggregations.py, https://github.com/xarray-contrib/flox/blob/8cfd999cf3d63a1f375e0447aaacb40ecacbcf9c/flox/aggregations.py#L84-L88
<details>
<summary>Full callstack</summary>

runpy.py,  line 198, in _run_module_as_main
  return _run_code(code, main_globals, None,
runpy.py,  line 88, in _run_code
  exec(code, run_globals)
test_core.py,  line 2266
  test_std_var_precision('nanvar','flox',10e10)
test_core.py,  line 2257, in test_std_var_precision
  no_offset, _ = groupby_reduce(array, labels, engine=engine, func=func)
core.py,  line 2865, in groupby_reduce
  results = _reduce_blockwise(
core.py,  line 1690, in _reduce_blockwise
  results = chunk_reduce(
core.py,  line 1311, in chunk_reduce
  result = generic_aggregate(
aggregations.py,  line 86, in generic_aggregate
  except AttributeError:
</details>

it's backending onto numpy and none of my new code (or the old broken code) is being tested. I'm guessing you'd need dask to evaluate the new chunkwise behaviour, but I'll hold off on implementing in case I've missed something where we can make the test work with pure numpy arrays

**Comment by dcherian** [comment]
I pushed a commit. Your changes are looking good! I constructed an expected result from numpy and it matches! I'm not sure what the expectation should be for `no_offset` vs `with_offset`. Does the original paper make some claims for this kind of comparison?

Lastly, I noticed that you basically have a ["property" test](https://www.youtube.com/watch?v=G0NUOst-53U) here (which is quite cool)

File: flox/core.py (hunk 7 above)
Comment (jemmajeffree) [reply flox/core.py:1293]: I'm partway through implementing something to work here. 

* How do I trigger this code pathway without brute force overwriting `if empty:` with `if True:`
* When np.full is called, `like` is a np array not a MultiArray, because it's (I think) the chunk data and bypassing var_chunk (could also be an artefact of the `if True` override above?). In a pinch, I guess I could add an elif that catches the `empty and reduction is var_chunk` and co-erce that into a MultiArray, but it's also ugly so I'm hoping you might have better ideas

**Comment by dcherian** [comment]
>      I'd expect it to fail for sufficiently big offsets, just to do better than the old algorithm. Does this present a problem?

Absolutely not! This is a massive improvement for common cases.

For now we can skip the failing comparison for large offsets. As long as it's close to numpy I'm happy. One thing to do would be to find the minimum tolerance for which we match numpy across that range of offsets.

File: tests/test_core.py [hunk 11]
```diff
@@ -2240,3 +2240,38 @@ def test_sparse_nan_fill_value_reductions(chunks, fill_value, shape, func):
...
+    expected = np.concatenate([np.nanvar(array[::2], keepdims=True), np.nanvar(array[1::2], keepdims=True)])
+    expected_offset = np.concatenate(
+        [np.nanvar(array[::2] + offset, keepdims=True), np.nanvar(array[1::2] + offset, keepdims=True)]
+    )
+
+    tol = {"rtol": 1e-8, "atol": 1e-10}  # Not sure how stringent to be here
+
+    assert_equal(expected, no_offset, tol)
+    assert_equal(expected_offset, with_offset, tol)
+    if exponent < 10:
```
Comment (jemmajeffree) [inline tests/test_core.py:2273]: If the line above is passing (`assert_equal(expected_offset, with_offset, tol)`) then it looks like the if statement shouldn't be needed here at all. There's no reason for the numpy implementation to be wrong in the same way as the new flox implementation

**Comment by jemmajeffree** [comment]
Does var_combine need a skipna too? (not just var_chunk?)

Happy to write this, just checking if my understanding is correct before I do so.

File: flox/aggregations.py [hunk 12]
```diff
@@ -355,7 +354,7 @@ def var_chunk(group_idx, array, *, engine: str, axis=-1, size=None, fill_value=N
         engine=engine,
         axis=axis,
         size=size,
-        fill_value=fill_value[2],  # Unpack fill value bc it's currently defined for multiarray
+        fill_value=0,  # Unpack fill value bc it's currently defined for multiarray
```
[the remark this answers is not in the document]
Comment (dcherian) [reply flox/aggregations.py:358]: Yes, I think the hardcoding is fine here. It's probably infe to just set `fill_value=(np.nan,)`

**Comment by dcherian** [comment]
> Does var_combine need a skipna too? (not just var_chunk?)

I would have thought so, but perhaps since the intermediates are using fill_value=0, the non-NaN skipping versions are fine.

**Comment by jemmajeffree** [comment]
found another problem :)
When I wrote var_combine, I assumed that the combine only called things in one direction at once
https://github.com/jemmajeffree/flox/blob/d5c59e35e748ef960c4a50bf5f46d188b86c9cfa/flox/aggregations.py#L408
`    assert len(axis) == 1, "Assuming that the combine function is only in one direction at once"` because the components need to be combined one at a time.

This assumption is broken with:

```python
import numpy as np
import xarray as xr
import flox

testdata = xr.DataArray(np.ones((4,2)),dims=('d0','d1')).chunk({'d0':2}).assign_coords({'g':xr.DataArray(np.arange(4)%2,dims='d0')})

testdata.groupby('g').var(('d0','d1')).load()
```

In the examples I've tried so far, only one of the axis dimensions actually has more than one value, so we could pick out that dimension for the cumsum/clip_first/clip_last? But I don't know what other situations might occur and whether that's actually 

Interesting that none of the automated tests picked this up, might write something to fix that when writing your other suggested fixes.

**Comment by dcherian** [comment]
Ah yes, we will need to add that

```
pytest "tests/test_core.py::test_cohorts_nd_by[flox-cohorts-None-nanvar]" --pdb
```

will drop you in a debugger prompt that you can use to explore the problem. we'll need to generalize this to nD. 

What will happens is that we combine the intermediates and concatenate them over the axes being reduced over. And then reduce over those axes. At the end, the result will have size =1 along those axes; which are then squeezed out (see `_squeeze_intermediates`).

**Comment by dcherian** [comment]
The property test failure indicates some kind of dtype behvaiour

```
import numpy as np
import dask.array

numpy_array = np.array([5592407., 5592407., 5592407.], dtype=np.float32)
by = np.array([0, 0, 0], dtype=np.int8)
array = dask.array.from_array(numpy_array, chunks=1)

np.var(numpy_array) # 0.25!!!!
np.var(numpy_array.compute())  # 0 

groupby_reduce(numpy_array, by, func="var", engine="numpy")[0] # 0.1111!
groupby_reduce(array, by, func="var", engine="numpy")[0].compute()  # 0.
```

Clearly we should be accumulating to float64 intermediates and then downcasting at the end if needed.

File: flox/aggregations.py [hunk 13]
```diff
@@ -427,7 +428,8 @@ def clip_first(array, n=1):
         + zero_denominator.astype(int)
     )
 
-    assert np.all((adj_terms * zero_denominator) == 0), (
+    check = adj_terms * zero_denominator
+    assert np.all(check[notnull(check)] == 0), (
```
Comment (dcherian) [inline flox/aggregations.py:432]: Is this right?

**Comment by jemmajeffree** [comment]
> Clearly we should be accumulating to float64 intermediates and then downcasting at the end if needed.

Unfortunately I think we need to turn into float64 _before_ calculating intermediates, such that the current dtypes parameter doesn't work? It's using [agg.dtypes['numpy']](https://github.com/xarray-contrib/flox/blob/34b551bd66f519f86393e22cde983c0916a5e119/flox/core.py#L1696) which is set to the final dtype [here?](https://github.com/xarray-contrib/flox/blob/34b551bd66f519f86393e22cde983c0916a5e119/flox/aggregations.py#L828)

The following commit _works_, but I don't like promoting to float64 as a final dtype

**Comment by jemmajeffree** [comment]
> Ah yes, we will need to add that
> 
> pytest "tests/test_core.py::test_cohorts_nd_by[flox-cohorts-None-nanvar]" --pdb
> will drop you in a debugger prompt that you can use to explore the problem. we'll need to generalize this to nD.
> 
> What will happens is that we combine the intermediates and concatenate them over the axes being reduced over. And then reduce over those axes. At the end, the result will have size =1 along those axes; which are then squeezed out (see _squeeze_intermediates).

Just to make sure I understand, there are situations in which var_combine needs to be able to handle combining intermediates along multiple axes at once? If so, that's tricky to handle because the equation I'm using merges MultiArray sets of intermediates pairwise. I think the best way to handle it is probably to stack those axes? (or is that bad for memory?) Or possibly deal with dimensions one after the other? Or we could do something awful with a 2D cumulative sum that loops around for calculating the adjustment terms, but that'd pretty much guarantee that the code is unintelligible to anybody else.

**Comment by jemmajeffree** [comment]
Thanks Deepak. I'm pretty busy the next few weeks, which is just to say if I drop off for a couple days it's not because I've lost interest. Back to normal 23rd September, but I'll try and get this across the line this week or early next.

I've got an sense of how to implement the reshape/multiple axes for var_chunk, and am hoping to get to this soon. I think this will fix most of the failing tests?

Regarding casting to float64, I'm not confident that I've thought through all the edge cases. ie, we probably wouldn't want to cast np.longdouble back to np.float64? Or is  it a safe assumption that np.float64 is a good idea for intermediates no matter what the inputs were? If you have a solution that you're happy with, then feel free to fix this one, otherwise I'll get to it when I can.

It's probably pretty obvious, but I've got basically no familiarity with pytest so I'm developing that familiarity while trying to use it here. I'm happy to keep working my way slowly through it, but I'm also happy for you to take on other failing tests. 

What do you see as the outstanding tasks to get this PR finalised? Is it just to address the causes of all the failing tests? I think we might have had a couple unresolved comment threads from code reviews that I'll try track down. Any suggestions for how you'd normally keep track of the last few things to do?

**Comment by jemmajeffree** [comment]
> Looks like you're just down to the property tests failing?! 

I think so? I wrapped the part of my code that combined along a single axis in a for loop that means it only has to handle one axis at a time. All the rest of the ideas I tried were nasty to implement.

> If so, I can clear those up.

Please :)

**Review by jemmajeffree — COMMENTED** [review]
Okay, fair.
I was deliberately trying to avoid triggering the "invalid value encountered in np divide" warning, though I concede it was a bit of a hack.
Do we want to force `den == 0` also to be NaN? (as opposed to inf?)

**Comment by dcherian** [comment]
so happy this is in! 👏🏾 👏🏾 👏🏾
